src/Cmfd.cpp

```diff
@@ -1750,22 +1774,278 @@ void Cmfd::setPolarQuadrature(PolarQuad* polar_quad) {
+ *         k is reduced to the number of neighbor cells for that instance.
+ */
+void Cmfd::generateKNearestStencils(){
+
+  std::vector< std::pair<int, FP_PRECISION> >::iterator iter2;
+  std::vector<int>::iterator iter;
+  Point* centroid;
+
+  /* Loop over mesh cells */
+  for (int i = 0; i < _num_x*_num_y; i++){
+
+    /* Loop over FRSs in mesh cell */
+    for (iter = _cell_fsrs.at(i).begin();
+         iter != _cell_fsrs.at(i).end(); ++iter) {
+
+      /* Get centroid */
+      centroid = _geometry->getFSRCentroid(*iter);
+
+      /* Create new stencil */
+      std::vector< std::pair<int, FP_PRECISION> > *stencil =
+        new std::vector< std::pair<int, FP_PRECISION> >;
+      _k_nearest_stencils[*iter] = (*stencil);
+
+      /* Get distance to all cells that touch current cell */
+      for (int j=0; j < 9; j++)
+        _k_nearest_stencils[*iter]
+          .push_back(std::make_pair<int, FP_PRECISION>
+                     (int(j), getDistanceToCentroid(centroid, i, j)));
+
+      /* Sort the distances */
+      std::sort(_k_nearest_stencils[*iter].begin(),
+                _k_nearest_stencils[*iter].end(), stencilCompare);
+
+      /* Remove non-existent cells */
+      iter2 = _k_nearest_stencils[*iter].begin();
+      while (iter2 != _k_nearest_stencils[*iter].end()){
+        if (iter2->second == std::numeric_limits<FP_PRECISION>::max())
+          iter2 = _k_nearest_stencils[*iter].erase(iter2++);
+        else
+          ++iter2;
+      }
+
+      /* Resize stencil to be of size <= _k_nearest */
+      _k_nearest_stencils[*iter].resize
+        (std::min(_k_nearest, int(_k_nearest_stencils[*iter].size())));
+    }
+  }
+}
+
+
+/**
+ * @brief Get the ratio used to update the FSR flux after converging CMFD.
+ * @detail This method takes in a cmfd cell, a MOC energy group, and a FSR
+ *         and returns the ratio used to update the FSR flux. There are two
+ *         methods that can be used to update the flux, conventional and
+ *         k-nearest centroid updating. The k-nearest centroid updating uses
+ *         the k-nearest cells (with k between 1 and 9) of the current CMFD
+ *         cell and the 8 neighboring CMFD cells. The stencil of cells
+ *         surrounding the current cell is defined as:
+ *
+ *                             6 7 8
+ *                             3 4 5
+ *                             0 1 2
+ *
+ *         where 4 is the given CMFD cell. If the cell is on the edge or corner
+ *         of the geometry and there are less than k nearest neighbor cells,
+ *         k is reduced to the number of neighbor cells for that instance.
+ * @param cmfd_cell The cmfd cell containing the FSR.
+ * @param moc_group The MOC energy group being updated.
+ * @param fsr The fsr being updated.
+ * @return the ratio used to update the FSR flux.
+ */
+FP_PRECISION Cmfd::getUpdateRatio(int cmfd_cell, int moc_group, int fsr){
+
+  FP_PRECISION ratio = 0.0;
+  FP_PRECISION total_distance = 1.e-10;
```

Comment:
Good question. Let's say we are using only one nearest and the FSR centroid is exactly in the middle of the CMFD cell. The `total_distance` will then be 0.0 and we'll get a divide by zero error when computing the ratio.